### PR TITLE
PYMT-1307 Provider aware search

### PR DESCRIPTION
**Pull request by ketanp77**
Search package now uses index name transformer to provision provider specific index names.

The `DefaultIndexTransformer` is the default index name transformer which keeps the behaviour as it was. Any custom index name transformation can be achieved via implementing `IndexTransformerInterface` and binding it in service container.

Ticket: [https://loyaltycorp.atlassian.net/browse/PYMT-1307](https://loyaltycorp.atlassian.net/browse/PYMT-1307)

### CODE REVIEW

File: src/Indexer.php [hunk 1]
```diff
@@ -60,7 +73,7 @@ public function clean(array $searchHandlers, ?bool $dryRun = null): IndexCleanRe
 
         /** @var \LoyaltyCorp\Search\Interfaces\SearchHandlerInterface[] $searchHandlers */
         foreach ($searchHandlers as $searchHandler) {
-            $handlerIndices[] = $searchHandler->getIndexName();
+            $handlerIndices = $this->indexTransformer->transformIndexNames($searchHandler);
```
Comment (merk) [inline src/Indexer.php:76]: You'll need to flatten this array?

File: src/Manager.php [hunk 2]
```diff
@@ -148,4 +145,21 @@ private function isHandled(EntitySearchHandlerInterface $handler, string $class)
...
+
+    /**
+     * Get index name.
+     *
+     * @param \LoyaltyCorp\Search\Interfaces\EntitySearchHandlerInterface $handler
+     * @param object $object
+     *
+     * @return string
+     */
+    private function getIndexName(EntitySearchHandlerInterface $handler, object $object): string
```
Comment (merk) [inline src/Manager.php]: This method kinda kills the seperation we've achieved with the TransformerInterface.

Could we add a method to that interface instead that receives the $object, index name and return the index name that should be used?

**Review by merk — CHANGES_REQUESTED** [review]

File: src/Indexer.php (hunk 1 above)
Comment (ketanp77) [reply src/Indexer.php:76]: Not really, `$handlerIndices` is still an array as default transformer returns array.

Default:
`['customers']`

Provider Specific:
`['customer_prov1', 'customer_prov2']`

File: src/Manager.php (hunk 2 above)
Comment (ketanp77) [reply src/Manager.php]: Ahh! forgot to move when refactoring.

**Review by merk — APPROVED** [review]

**Pull request merged**